lookup_estim.go

```diff
@@ -0,0 +1,302 @@
+	}
+
+	// count number of peers we have already (successfully) contacted via the above method
+	sentAndSuccessCount := 0
+	for _, s := range es.peerStates {
+		if s == Sent || s == Success {
+			sentAndSuccessCount += 1
+		}
+	}
+
+	// if we have already contacted more than bucketSize peers stop the procedure
+	if sentAndSuccessCount >= es.dht.bucketSize {
+		return true
+	}
+
+	// calculate average distance of the set of closest peers
+	sum := 0.0
+	for _, d := range distances {
+		sum += d
+	}
+	avg := sum / float64(len(distances))
+
+	// if the average is below the set threshold stop the procedure
+	return avg < es.setThreshold
+}
+
+func (es *estimatorState) putProviderRecord(pid peer.ID) {
+	err := es.dht.protoMessenger.PutProvider(es.putCtx, pid, []byte(es.key), es.dht.host)
+	es.peerStatesLk.Lock()
+	if err != nil {
+		es.peerStates[pid] = Failure
+	} else {
+		es.peerStates[pid] = Success
+	}
+	es.peerStatesLk.Unlock()
+
+	// indicate that this ADD_PROVIDER RPC has completed
+	es.doneChan <- struct{}{}
+}
+
+// waitForRPCs waits for a subset of ADD_PROVIDER RPCs to complete and then acquire a lease on
+// a bound channel to return early back to the user and prevent unbound asynchronicity. If
+// there are already too many requests in-flight we are just waiting for our current set to
+// finish.
+func (es *estimatorState) waitForRPCs() {
+	es.peerStatesLk.RLock()
+	rpcCount := len(es.peerStates)
+	es.peerStatesLk.RUnlock()
+
+	// returnThreshold can't be larger than the total number issued RPCs
+	if es.returnThreshold > rpcCount {
+		es.returnThreshold = rpcCount
+	}
+
+	// Wait until returnThreshold ADD_PROVIDER RPCs have returned
+	for range es.doneChan {
+		if int(es.putProvDone.Add(1)) == es.returnThreshold {
+			break
+		}
+	}
+	// At this point only a subset of all ADD_PROVIDER RPCs have completed.
+	// We want to give control back to the user as soon as possible because
+	// it is highly likely that at least one of the remaining RPCs will time
+	// out and thus slow down the whole processes. The provider records will
+	// already be available with less than the total number of RPCs having
+	// finished. This has been investigated here:
+	// https://github.com/protocol/network-measurements/blob/master/results/rfm17-provider-record-liveness.md
+
+	// For the remaining ADD_PROVIDER RPCs try to acquire a lease on the optProvJobsPool channel.
+	// If that worked we need to consume the doneChan and release the acquired lease on the
+	// optProvJobsPool channel.
+	remaining := rpcCount - int(es.putProvDone.Load())
+	for i := 0; i < remaining; i++ {
+		select {
+		case es.dht.optProvJobsPool <- struct{}{}:
+			// We were able to acquire a lease on the optProvJobsPool channel.
+			// Consume doneChan to release the acquired lease again.
+			go es.consumeDoneChan(rpcCount)
+		case <-es.doneChan:
+			// We were not able to acquire a lease but an ADD_PROVIDER RPC resolved.
+			if int(es.putProvDone.Add(1)) == rpcCount {
```

Comment:
do the inverse, set `es.PutProvDone` to `rpcCount` at the start, and have workers do:
```go
var minusOne uint32
minusOne-- // go doesn't want to constant cast -1 to an uint32, that does the same thing, this also gets folded.
if es.putProvDone.Add(minusOne) == 0  {
```
Also I guess `putProvDone` should be renamed `putProvInflight`

Reply:
I think the challenge with both your suggestions is that we don't know ahead of time if we are going to acquire a lease for an inflight `putProviderRecord` call. We would need to inflight communicate to it that we acquired a lease for it so that it'll release all resources for us.

I think I see where you want to go with both of your suggestions but I can't see a simpler way to do that.